core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -394,17 +394,22 @@ void connect(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
 
         final Bootstrap bootstrap = getBootstrap(desiredProtocol);
 
-        final Channel channel = bootstrap.register().channel();
-        configureProxy(channel, poolKey.proxyConfig, desiredProtocol);
-        final ChannelFuture connectFuture = channel.connect(remoteAddress);
-
-        connectFuture.addListener((ChannelFuture future) -> {
-            if (future.isSuccess()) {
-                initSession(desiredProtocol, poolKey, future, sessionPromise);
-            } else {
-                invokeProxyConnectFailed(desiredProtocol, poolKey, future.cause());
-                sessionPromise.tryFailure(future.cause());
+        bootstrap.register().addListener((ChannelFuture registerFuture) -> {
+            if (!registerFuture.isSuccess()) {
+                sessionPromise.tryFailure(registerFuture.cause());
+                return;
             }
+
+            final Channel channel = registerFuture.channel();
+            configureProxy(channel, poolKey.proxyConfig, desiredProtocol);
```

Comment:
Is there any chance where `configureProxy()` raises an exception? We may have to do this for safety:

```java
try {
    configureProxy(...);
    channel.connect(...).addListener(
        ...
    );
} catch (Throwable cause) {
    invokeProxyConnectFailed(desiredProtocol, poolKey, cause);
    sessionPromise.tryFailure(connectFuture.cause());
}
```

Reply:
Updated, looks like not happen now, but better to wrap to avoid future error?

Reply:
That was what I meant :smile: